key.go

```diff
@@ -21,16 +22,30 @@ func BinaryFromDsKey(k datastore.Key) ([]byte, error) {
 	return base32.RawStdEncoding.DecodeString(k.String()[1:])
 }
 
+// MultihashToDsKey creates a Key from the given Multihash.
+func MultihashToDsKey(k mh.Multihash) datastore.Key {
+	return NewKeyFromBinary(k)
+}
+
+// DsKeyToMultihash converts a dsKey to the corresponding Multihash.
+func DsKeyToMultihash(dsKey datastore.Key) (mh.Multihash, error) {
+	kb, err := BinaryFromDsKey(dsKey)
+	if err != nil {
+		return nil, err
+	}
+	return mh.Cast(kb)
+}
+
 // CidToDsKey creates a Key from the given Cid.
 func CidToDsKey(k cid.Cid) datastore.Key {
-	return NewKeyFromBinary(k.Bytes())
+	return MultihashToDsKey(k.Hash())
 }
 
 // DsKeyToCid converts the given Key to its corresponding Cid.
 func DsKeyToCid(dsKey datastore.Key) (cid.Cid, error) {
-	kb, err := BinaryFromDsKey(dsKey)
+	hash, err := DsKeyToMultihash(dsKey)
 	if err != nil {
-		return cid.Cid{}, err
+		return cid.Cid{}, nil
```

Comment:
@Stebalien why did you return nil?

Reply:
I have absolutely no idea. That's really not good.